<html>
<head>
<title>
Genometric Correlation project.
</title>
<style type="text/css">
<!--

.content {
	margin: 40px 40px 40px 40px;
	font-family: Verdana, Arial, Helvetica, sans-serif;
	font-size: 18px;
	line-height: 24px;
	color: #000000;
}

.overview {
	background-color: #FEF1B5;
	padding: 10px;
}

h1 {
	font-family: Verdana, Arial, Helvetica, sans-serif;
	font-size: 36px;
	font-weight: bold;
	line-height: 48px;
	color: #8B5A00;
}

-->
</style>
<link rel="shortcut icon" href="gc_logo.ico"/>
</head>
<body class="content">
<h1>GenometriCorr USCS examples</h1>
<p class="overview">
We used two BED files in examples, both are generated by <a href=http://genome.ucsc.edu/>USCS genome bioinformatics browser</a>.
</p>
<p>

The coords of USCS CpG islands were retrieved by <a href=http://genome.ucsc.edu/cgi-bin/hgTables?clade=mammal&command=start&db=hg19&hgta_database=hg19&hgta_fieldSelectTable=hg19.cpgIslandExt&hgta_fs.check.hg19.cpgIslandExt.chrom=1&hgta_fs.check.hg19.cpgIslandExt.chromEnd=1&hgta_fs.check.hg19.cpgIslandExt.chromStart=1&gta_fs.check.hg19.cpgIslandExt.cpgNum=0&hgta_fs.check.hg19.cpgIslandExt.gcNum=0&hgta_fs.check.hg19.cpgIslandExt.length=0&hgta_fs.check.hg19.cpgIslandExt.name=0&hgta_fs.check.hg19.cpgIslandExt.obsExp=0&hgta_fs.check.hg19.cpgIslandExt.perCpg=0&hgta_fs.check.hg19.cpgIslandExt.perGc=0&hgta_group=regulation&hgta_outputType=primaryTable&hgta_regionType=genome&hgta_table=cpgIslandExt&hgta_track=cpgIslandExt&hgta_doPrintSelectedFields=&org=Human&>http://genome.ucsc.edu/cgi-bin/hgTables?clade=mammal&command=start&db=hg19&hgta_database=hg19&hgta_fieldSelectTable=hg19.cpgIslandExt&hgta_fs.check.hg19.cpgIslandExt.chrom=1&hgta_fs.check.hg19.cpgIslandExt.chromEnd=1&hgta_fs.check.hg19.cpgIslandExt.chromStart=1&gta_fs.check.hg19.cpgIslandExt.cpgNum=0&hgta_fs.check.hg19.cpgIslandExt.gcNum=0&hgta_fs.check.hg19.cpgIslandExt.length=0&hgta_fs.check.hg19.cpgIslandExt.name=0&hgta_fs.check.hg19.cpgIslandExt.obsExp=0&hgta_fs.check.hg19.cpgIslandExt.perCpg=0&hgta_fs.check.hg19.cpgIslandExt.perGc=0&hgta_group=regulation&hgta_outputType=primaryTable&hgta_regionType=genome&hgta_table=cpgIslandExt&hgta_track=cpgIslandExt&hgta_doPrintSelectedFields=&org=Human&</a>.

</p>
<p>
The coords of RefSeq genes were retrieved by <a href=http://genome.ucsc.edu/cgi-bin/hgTables?db=hg19&hgta_database=hg19&hgta_group=genes&hgta_track=refGene&hgta_table=refGene&hgta_regionType=genome&hgta_outputType=primaryTable&hgta_fieldSelectTable=hg19.refGene&hgta_fs.check.hg19.refGene.chrom=1&hgta_fs.check.hg19.refGene.name=0&hgta_fs.check.hg19.refGene.txEnd=1&hgta_fs.check.hg19.refGene.txStart=1&hgta_doPrintSelectedFields=&>http://genome.ucsc.edu/cgi-bin/hgTables?db=hg19&hgta_database=hg19&hgta_group=genes&hgta_track=refGene&hgta_table=refGene&hgta_regionType=genome&hgta_outputType=primaryTable&hgta_fieldSelectTable=hg19.refGene&hgta_fs.check.hg19.refGene.chrom=1&hgta_fs.check.hg19.refGene.name=0&hgta_fs.check.hg19.refGene.txEnd=1&hgta_fs.check.hg19.refGene.txStart=1&hgta_doPrintSelectedFields=&</a>.

</p>
<p>
Creation of such URLs is described here <a href=http://genomewiki.ucsc.edu/index.php/Table_Browser_URL>GenomeWiki</a> and here <a href=http://digitheadslabnotebook.blogspot.com/2009/02/spelunking-in-ucsc-genome-browser.html>DigitHead's lab notebook</a>.
</p> 
<p>
The chromosome length data was extracted from the <a href=http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz>UCSC hg19 chromosome information table</a>.
</body>
</html>


